### PR TITLE
pubsub: acquire FlowController before releasing

### PR DESCRIPTION
**Pull request by pongad**

### CODE REVIEW

File: google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/spi/v1/MessageDispatcher.java [hunk 1]
```diff
@@ -260,8 +260,7 @@ public int getMessageDeadlineSeconds() {
   }
 
   public void processReceivedMessages(List<com.google.pubsub.v1.ReceivedMessage> responseMessages) {
-    int receivedMessagesCount = responseMessages.size();
-    if (receivedMessagesCount == 0) {
+    if (responseMessages.size() == 0) {
```
Comment (garrettjonesgoogle) [inline google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/spi/v1/MessageDispatcher.java]: @pongad this seems worth doing.

File: google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/spi/v1/MessageDispatcher.java [hunk 2]
```diff
@@ -276,7 +275,13 @@ public void processReceivedMessages(List<com.google.pubsub.v1.ReceivedMessage> r
     logger.log(
         Level.FINER, "Received {0} messages at {1}", new Object[] {responseMessages.size(), now});
 
+    try {
+      flowController.reserve(responseMessages.size(), totalByteCount);
+    } catch (FlowController.FlowControlException unexpectedException) {
+      throw new IllegalStateException("Flow control unexpected exception", unexpectedException);
```
Comment (garrettjonesgoogle) [inline google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/spi/v1/MessageDispatcher.java]: Just use `e` for the exception variable. (Ok, I realize this existed before, it it seems odd.)

Also, why is this an `IllegalStateException`?

File: google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/spi/v1/MessageDispatcher.java [hunk 3]
```diff
@@ -276,7 +275,13 @@ public void processReceivedMessages(List<com.google.pubsub.v1.ReceivedMessage> r
...
+    }
     messagesWaiter.incrementPendingMessages(responseMessages.size());
+
     Iterator<AckHandler> acksIterator = ackHandlers.iterator();
```
Comment (garrettjonesgoogle) [inline google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/spi/v1/MessageDispatcher.java]: I know this is a request to change code outside of the scope of your fix, but your fix brought to my attention that it's a little hard to follow this method because there are some intermingled things. In particular, the loop to count the total bytes also creates the ack handlers. While this avoids two loops, it also makes the process much more coupled and harder to reason through. Could we pull out the byte counting and flow control reservation into a separate method?

File: google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/spi/v1/MessageDispatcher.java (hunk 2 above)
Comment (pongad) [inline google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/spi/v1/MessageDispatcher.java]: Renamed.

Dug through some history. FlowController used to have its parameters split: the numerics (how many messages and bytes) and whether to throw exception or stall. The user used to only be able to control the first part. We unconditionally tell the FlowController to stall and never throws. We can therefore "never" see this exception.

Now that we have changed FlowControlSettings to also contain the LimitExceededBehavior, this is no longer true. I guess we can check at construction-time that the behavior is always IGNORE or BLOCK...? WDYT?

File: google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/spi/v1/MessageDispatcher.java (hunk 1 above)
Comment (pongad) [inline google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/spi/v1/MessageDispatcher.java]: Done

File: google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/spi/v1/MessageDispatcher.java (hunk 3 above)
Comment (pongad) [inline google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/spi/v1/MessageDispatcher.java]: Done

**Comment by pongad** [comment]
Going through this the second revealed more bugs. I shuffled around a couple more things and documented some happen-before edges that we need to observe.

@garrettjonesgoogle PTAL
@davidtorres could you please take a look?

File: google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/spi/v1/MessageDispatcher.java [hunk 4]
```diff
@@ -307,12 +326,14 @@ public void run() {
             }
           });
     }
+  }
 
-    synchronized (outstandingAckHandlers) {
-      outstandingAckHandlers.add(
-          new ExtensionJob(expiration, INITIAL_ACK_DEADLINE_EXTENSION_SECONDS, ackHandlers));
+  private static int totalMessageSize(Collection<ReceivedMessage> messages)  {
```
Comment (garrettjonesgoogle) [inline google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/spi/v1/MessageDispatcher.java]: rename to `getTotalMessageSize`

File: google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/spi/v1/MessageDispatcher.java [hunk 5]
```diff
@@ -260,28 +261,46 @@ public int getMessageDeadlineSeconds() {
...
     }
-    Instant now = new Instant(clock.millisTime());
-    int totalByteCount = 0;
+
     final ArrayList<AckHandler> ackHandlers = new ArrayList<>(responseMessages.size());
     for (ReceivedMessage pubsubMessage : responseMessages) {
-      int messageSize = pubsubMessage.getMessage().getSerializedSize();
-      totalByteCount += messageSize;
-      ackHandlers.add(new AckHandler(pubsubMessage.getAckId(), messageSize));
+      ackHandlers.add(new AckHandler(pubsubMessage.getAckId(), pubsubMessage.getMessage().getSerializedSize()));
```
Comment (garrettjonesgoogle) [inline google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/spi/v1/MessageDispatcher.java]: Could you create the new `AckHandler` on its own line to make the lines less complex?

File: google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/spi/v1/MessageDispatcher.java [hunk 6]
```diff
@@ -260,28 +261,46 @@ public int getMessageDeadlineSeconds() {
...
+    Instant now = new Instant(clock.millisTime());
     Instant expiration = now.plus(messageDeadlineSeconds * 1000);
     logger.log(
         Level.FINER, "Received {0} messages at {1}", new Object[] {responseMessages.size(), now});
 
+    synchronized (outstandingAckHandlers) {
+      // AckDeadlineAlarm modifies lists in outstandingAckHandlers in-place and might run at any time.
+      // We will also later iterate over ackHandlers when we give messages to user code.
+      // We must create a new list to pass to outstandingAckHandlers,
+      // so that we can't iterate and modify the list concurrently.
```
Comment (garrettjonesgoogle) [inline google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/spi/v1/MessageDispatcher.java]: If you want to call out the copy of `ackHandlers`, the pull out the creation of the list copy to its own line before the `add` call.

File: google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/spi/v1/MessageDispatcher.java [hunk 7]
```diff
@@ -260,28 +261,46 @@ public int getMessageDeadlineSeconds() {
...
-    } catch (FlowController.FlowControlException unexpectedException) {
-      throw new IllegalStateException("Flow control unexpected exception", unexpectedException);
+      flowController.reserve(responseMessages.size(), totalMessageSize(responseMessages));
+    } catch (FlowController.FlowControlException e) {
+      throw new IllegalStateException("Flow control unexpected exception", e);
     }
     messagesWaiter.incrementPendingMessages(responseMessages.size());
 
+    // Reserving flow control must happen before we give the messages to the user,
+    // otherwise the user code might be given too many messages to process at once.
```
Comment (garrettjonesgoogle) [inline google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/spi/v1/MessageDispatcher.java]: Put the comment about reserving flow control before the flow control call, not after

File: google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/spi/v1/MessageDispatcher.java [hunk 8]
```diff
@@ -260,28 +261,46 @@ public int getMessageDeadlineSeconds() {
...
+      outstandingAckHandlers.add(
+          new ExtensionJob(
+              expiration,
+              INITIAL_ACK_DEADLINE_EXTENSION_SECONDS,
+              new ArrayList<AckHandler>(ackHandlers)));
+    }
+    setupNextAckDeadlineExtensionAlarm(expiration);
+
+    // Deadline extension must be set up before we reserve flow control.
+    // Flow control might block for a while, and extension will keep messages from expiring.
```
Comment (garrettjonesgoogle) [inline google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/spi/v1/MessageDispatcher.java]: Put the comment about extension before the extension code, not after.

File: google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/spi/v1/MessageDispatcher.java (hunk 4 above)
Comment (pongad) [inline google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/spi/v1/MessageDispatcher.java]: Done

File: google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/spi/v1/MessageDispatcher.java (hunk 7 above)
Comment (pongad) [inline google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/spi/v1/MessageDispatcher.java]: Done

File: google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/spi/v1/MessageDispatcher.java (hunk 8 above)
Comment (pongad) [inline google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/spi/v1/MessageDispatcher.java]: Done

File: google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/spi/v1/MessageDispatcher.java (hunk 6 above)
Comment (pongad) [inline google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/spi/v1/MessageDispatcher.java]: Done

File: google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/spi/v1/MessageDispatcher.java (hunk 5 above)
Comment (pongad) [inline google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/spi/v1/MessageDispatcher.java]: Done

**Comment by pongad** [comment]
@garrettjonesgoogle Thank you for the rigorous review. PTAL.
